### PR TITLE
Silent subcommands

### PR DESCRIPTION
**Pull request by phlptp**
add a silent option to subcommands to prevent the use from showing up in the subcommands list if used.

This is in response to Issue #522.

Adds an ability to silence subcommands and prevent them from showing up in the used subcommands list.
This allows things like

```c++
    auto sub1 = app.add_subcommand("help")->silent();
    sub1->parse_complete_callback([]() { throw CLI::CallForHelp(); });
```

### CODE REVIEW

File: book/chapters/subcommands.md [hunk 1]
```diff
@@ -112,3 +112,18 @@ Here, `--shared_flag` was set on the main app, and on the command line it "falls
 
 This is a special mode that allows "prefix" commands, where the parsing completely stops when it gets to an unknown option. Further unknown options are ignored, even if they could match. Git is the traditional example for prefix commands; if you run git with an unknown subcommand, like "`git thing`", it then calls another command called "`git-thing`" with the remaining options intact.
 
+### Silent subcommands
+
+Subcommands can be modified by using the `silent` option.  This will prevent the subcommand from showing up in the get_subcommands list.  This can be used to make subcommands into modifiers For example a help subcommand might look like
```
Comment (henryiii) [inline book/chapters/subcommands.md]:
```suggestion
Subcommands can be modified by using the `silent` option.  This will prevent the subcommand from showing up in the get_subcommands list.  This can be used to make subcommands into modifiers. For example, a help subcommand might look like
```

File: include/CLI/App.hpp [hunk 2]
```diff
@@ -222,7 +222,11 @@ class App {
     App *parent_{nullptr};
 
     /// Counts the number of times this command/subcommand was parsed
-    std::size_t parsed_{0};
+    std::uint32_t parsed_{0U};
+
+    /// indicator that the subcommand is silent and won't show up in subcommands list
+    /// This is potentially useful as a modifier subcommand
+    bool silent_{false};
```
Comment (henryiii) [inline include/CLI/App.hpp:223]: This should be moved up to be with the other 8-bit members; if bools are placed together the compiler is more likely to pack them into 8 bits each, rather than adding alignment spacing. Technically, we probably should try to order the struct members large to small, but optimizing this is not really that important - it's just nice to do it right when we add something.

File: include/CLI/App.hpp [hunk 3]
```diff
@@ -2390,8 +2404,8 @@ class App {
     ///
     /// If this has more than one dot.separated.name, go into the subcommand matching it
     /// Returns true if it managed to find the option, if false you'll need to remove the arg manually.
-    void _parse_config(const std::vector<ConfigItem> &args) {
-        for(const ConfigItem &item : args) {
+    void _parse_config(std::vector<ConfigItem> &args) {
+        for(ConfigItem item : args) {
```
Comment (henryiii) [inline include/CLI/App.hpp]: Why this change?

Comment (phlptp) [reply include/CLI/App.hpp]: recommended by clang-tidy or some other static analysis tools I ran

File: include/CLI/App.hpp (hunk 2 above)
Comment (phlptp) [reply include/CLI/App.hpp:223]: I was trying to put it into a gap in alignment, I rearranged a bit

File: include/CLI/App.hpp (hunk 3 above)
Comment (phlptp) [reply include/CLI/App.hpp]: I am going to change it back, it might have been a merge conflict or something

**Comment by phlptp** [comment]
I think this is ready now @henryiii 
I also upgraded a version number with the cuda build for a github change

**Review by henryiii — APPROVED** [review]
